### PR TITLE
Add map component

### PR DESCRIPTION
**Pull request by codemayank**
* **Please check if the PR fulfills these requirements**
- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Feature

* **What is the current behavior?** (You can also link to an open issue here)
Addresses #275. Adds Map component.

* **What is the new behavior (if this is a feature change)?**
Map component maps over an array to output idyll components. e.g the output of below will be three Display components with values one, two and three.
```
[Map value:`['one', 'two', 'three]` currentValue:'iterator']
[Display value:iterator' /]
[/Map]
```

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No

* **Other information**:

### CODE REVIEW

File: packages/idyll-components/src/map.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+          let newProps = Object.assign({}, child.props);
+          newProps = Object.keys(child.props).reduce((props, elm) => {
+            if (props[elm] === currentValue) {
+              props[elm] = val;
+              return props;
+            }
+            return props;
+          }, newProps);
+          return (
+            <TextContainer>
```
Comment (mathisonian) [inline packages/idyll-components/src/map.js]: I don't think we need this `TextContainer` here

Comment (codemayank) [reply packages/idyll-components/src/map.js]: I have updated this to remove the TextContainer.

File: packages/idyll-components/src/map.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    if (children) {
+      return mapChildren(children, child => {
+        return value.map(val => {
+          let newProps = Object.assign({}, child.props);
+          newProps = Object.keys(child.props).reduce((props, elm) => {
+            if (props[elm] === currentValue) {
+              props[elm] = val;
+              return props;
+            }
```
Comment (mathisonian) [inline packages/idyll-components/src/map.js]: Let's remove the references to `currentValue` and just have a hard coded keyword that will be replaced by the mapper, say, `_iterator`.

That way

```
[Map value:`['one', 'two', 'three]`]
  [Display value:'_iterator' /]
[/Map]
```

will work, and I think its a little more straightforward.

**Comment by mathisonian** [comment]
What do you think about renaming to `[Loop]`?

/cc @tanalan @megan-vo

**Comment by codemayank** [comment]
> What do you think about renaming to `[Loop]`?

`[Loop]` would be a good name for this. I'll push the final changes once everyone agrees on the name.

**Comment by tanalan** [comment]
> What do you think about renaming to `[Loop]`?
> 
> /cc @tanalan @megan-vo

`Loop` sounds good to me!

**Comment by mathisonian** [comment]
> [Loop] would be a good name for this. I'll push the final changes once everyone agrees on the name.

Sounds good 👍 let's go with `[Loop]` and I'll merge once the final updates are in

File: packages/idyll-components/src/map.js (hunk 2 above)
Comment (codemayank) [reply packages/idyll-components/src/map.js]: Updated this as per your comment.

**Comment by codemayank** [comment]
@mathisonian I have pushed the final changes.

**Comment by mathisonian** [comment]
Thanks @codemayank! ✨

**Pull request merged**